src/api/writer-generator/typescript.ts

```diff
@@ -508,8 +613,859 @@ export class TypeScript extends Writer<TypeScriptOptions> {
+        if (!isSpecializationTypeSchema(specialization)) return overrides;
+
+        for (const [fieldName, pField] of Object.entries(flatProfile.fields ?? {})) {
+            if (!isNotChoiceDeclarationField(pField)) continue;
+            const sField = specialization.fields?.[fieldName];
+            if (!sField || isChoiceDeclarationField(sField)) continue;
+
+            // Check for Reference narrowing
+            if (pField.reference && sField.reference && pField.reference.length < sField.reference.length) {
+                const references = pField.reference
+                    .map((ref) => {
+                        const resRef = tsIndex.findLastSpecializationByIdentifier(ref);
+                        if (resRef.name !== ref.name) {
+                            return `"${resRef.name}"`;
+                        }
+                        return `"${ref.name}"`;
+                    })
+                    .join(" | ");
+                overrides.set(fieldName, {
+                    profileType: `Reference<${references}>`,
+                    required: pField.required ?? false,
+                    array: pField.array ?? false,
+                });
+            }
+            // Check for cardinality change (optional -> required)
+            else if (pField.required && !sField.required) {
+                const tsType = this.tsTypeForProfileField(tsIndex, flatProfile, fieldName, pField);
+                overrides.set(fieldName, {
+                    profileType: tsType,
+                    required: true,
+                    array: pField.array ?? false,
+                });
+            }
+        }
+        return overrides;
+    }
+
+    /**
+     * Generates an override interface for profiles that narrow cardinality or Reference types.
+     * Example: export interface USCorePatient extends Patient { subject: Reference<"Patient"> }
+     */
+    generateProfileOverrideInterface(tsIndex: TypeSchemaIndex, flatProfile: ProfileTypeSchema) {
+        const overrides = this.detectFieldOverrides(tsIndex, flatProfile);
+        if (overrides.size === 0) return;
+
+        const tsProfileName = tsResourceName(flatProfile.identifier);
+        const tsBaseResourceName = tsResourceName(flatProfile.base);
+
+        this.curlyBlock(["export", "interface", tsProfileName, "extends", tsBaseResourceName], () => {
+            for (const [fieldName, override] of overrides) {
+                const tsField = tsFieldName(fieldName);
+                const optionalSymbol = override.required ? "" : "?";
+                const arraySymbol = override.array ? "[]" : "";
+                this.lineSM(`${tsField}${optionalSymbol}: ${override.profileType}${arraySymbol}`);
+            }
+        });
+        this.line();
+    }
+
     generateResourceModule(tsIndex: TypeSchemaIndex, schema: TypeSchema) {
-        this.cat(`${tsModuleFileName(schema.identifier)}`, () => {
+        const generateModule = () => {
```

Comment:
Looks like if we inline statements into the next if -- code will be much simpler.